tests/templates/test_subroutines/test_prepselprep.py

```diff
@@ -0,0 +1,393 @@
+        assert all(qml.equal(op1, op2) for op1, op2 in zip(op.ops, new_op.ops))
+        assert op.control == new_op.control
+        assert op.wires == new_op.wires
+        assert op.target_wires == new_op.target_wires
+        assert op is not new_op
+
+
+def test_control_in_ops():
+    """Test that using an operation wire as a control wire results in an error"""
+
+    lcu = qml.dot([1 / 2, 1 / 2], [qml.Identity(0), qml.PauliZ(0)])
+    with pytest.raises(ValueError, match="Control wires should be different from operation wires."):
+        qml.PrepSelPrep(lcu, control=0)
+
+
+class TestInterfaces:
+    """Tests that the template is compatible with interfaces used to compute gradients"""
+
+    @pytest.mark.autograd
+    def test_autograd(self):
+        """Test the autograd interface"""
+
+        dev = qml.device("default.qubit")
+
+        coeffs = pnp.array([1 / 2, 1 / 2], requires_grad=True)
+        ops = [qml.Identity(1), qml.PauliZ(1)]
+
+        @qml.qnode(dev, diff_method="finite-diff")
+        def prepselprep(coeffs):
+            lcu = qml.ops.LinearCombination(coeffs, ops)
+            qml.PrepSelPrep(lcu, control=0)
+            return qml.expval(qml.Z(1))
+
+        @qml.qnode(dev)
+        def manual(coeffs):
+            normalized_coeffs = qml.math.sqrt(coeffs) / qml.math.norm(qml.math.sqrt(coeffs))
+
+            qml.StatePrep(normalized_coeffs, wires=0)
+            qml.Select(ops, control=0)
+            qml.adjoint(qml.StatePrep(normalized_coeffs, wires=0))
+            return qml.expval(qml.Z(1))
+
+        grad_prepselprep = qml.grad(prepselprep)(coeffs)
+        grad_manual = qml.grad(manual)(coeffs)
+
+        assert qml.math.allclose(grad_prepselprep, grad_manual)
+
+        fd_prepselprep = qml.jacobian(prepselprep)(coeffs)
+        assert qml.math.allclose(grad_prepselprep, fd_prepselprep)
+
+    @pytest.mark.jax
+    def test_jax(self):
+        """Test the jax interface"""
+        import jax
+        import jax.numpy as jnp
+
+        dev = qml.device("default.qubit")
+        coeffs = jnp.array([1 / 2, 1 / 2])
+        ops = [qml.Identity(1), qml.PauliZ(1)]
+
+        @qml.qnode(dev)
+        def prepselprep(coeffs):
+            lcu = qml.ops.LinearCombination(coeffs, ops)
+            qml.PrepSelPrep(lcu, control=0)
+            return qml.expval(qml.Z(1))
+
+        @qml.qnode(dev)
+        def manual(coeffs):
+            normalized_coeffs = qml.math.sqrt(coeffs) / qml.math.norm(qml.math.sqrt(coeffs))
+
+            qml.StatePrep(normalized_coeffs, wires=0)
+            qml.Select(ops, control=0)
+            qml.adjoint(qml.StatePrep(normalized_coeffs, wires=0))
+            return qml.expval(qml.Z(1))
+
+        grad_prepselprep = jax.grad(prepselprep)(coeffs)
+        grad_manual = jax.grad(manual)(coeffs)
+
+        assert qml.math.allclose(grad_prepselprep, grad_manual)
+
+    @pytest.mark.jax
+    def test_jax_jit(self):
```

Comment:
Adding `jax.jit` in this example does not seem to work. 
Why do you think this error is currently popping up?


```
import pennylane as qml
import jax


coeffs = jax.numpy.array([-1., 1., 1.])
ops = [qml.Z(1), qml.X(1), qml.Y(1)]
dev = qml.device('default.qubit')

@jax.jit
@qml.qnode(dev)
def circuit(coeffs):

    lcu = qml.ops.LinearCombination(coeffs, ops)
    qml.PrepSelPrep(lcu, control = [0,2])  
    return qml.state()

circuit(coeffs)
```

Could you figure out how to fix it?

Reply:
It's because of the control flow used to handle negative and complex coefficients in the ```compute_decomposition``` function. I changed it to this and the error went away.

```
        for coeff, op in zip(*lcu.terms()):
            real = qml.math.real(coeff)
            sign = qml.math.sign(real)
            new_coeffs.append(sign*real)
            new_op = qml.ops.LinearCombination([sign], [op])
            new_ops.append(new_op)

            imag = qml.math.imag(coeff)
            sign = qml.math.sign(imag)
            new_coeffs.append(sign*imag)
            new_op = qml.ops.LinearCombination([1j*sign], [op])
            new_ops.append(new_op)
```

However, this has its own problems. This adds the zero matrix to the LCU, which seems problematic as it is not unitary. It seems the controlled version of these operations do not get queued after calling this.

```
        with qml.QueuingManager.stop_recording():
            prep_ops = qml.StatePrep.compute_decomposition(normalized_coeffs, control)
            select_ops = qml.Select.compute_decomposition(new_ops, control)
            adjoint_prep_ops = qml.adjoint(
                qml.StatePrep(normalized_coeffs, control)
            ).decomposition()

        ops = prep_ops + select_ops + adjoint_prep_ops

        for op in ops:
            if qml.QueuingManager.recording():
                qml.apply(op)
```

Reply:
I've resolved the issues. I can't get JIT to work with the preprocessing that handles negative/complex coefficients or LCUs with a non-power of 2 number of terms. The preprocessing involves modifying input in a way that can't be done with JAX's tracer objects.

As a temporary fix I added an optional `jit` flag to `qml.PrepSelPrep`. The flag lets you skip the preprocessing steps, so it assumes the LCU is the right form.